test/e2e/clusterclass_changes.go

```diff
@@ -464,6 +590,40 @@ func assertMachineDeploymentTopologyFields(g Gomega, md clusterv1.MachineDeploym
 	}
 }
 
+// assertMachinePoolTopologyFields asserts that all fields set in the MachinePoolTopology have been set on the MachinePool.
+// Note: We intentionally focus on the fields set in the MachinePoolTopology and ignore the ones set through ClusterClass
+// as we want to validate that the fields of the MachinePoolTopology have been propagated correctly.
+func assertMachinePoolTopologyFields(g Gomega, mp expv1.MachinePool, mpTopology clusterv1.MachinePoolTopology) {
+	// Note: We only verify that all labels and annotations from the Cluster topology exist to keep it simple here.
+	// This is fully covered by the ClusterClass rollout test.
+	for k, v := range mpTopology.Metadata.Labels {
+		g.Expect(mp.Labels).To(HaveKeyWithValue(k, v))
+	}
+	for k, v := range mpTopology.Metadata.Annotations {
+		g.Expect(mp.Annotations).To(HaveKeyWithValue(k, v))
+	}
+
+	if mpTopology.NodeDrainTimeout != nil {
+		g.Expect(mp.Spec.Template.Spec.NodeDrainTimeout).To(Equal(mpTopology.NodeDrainTimeout))
+	}
+
+	if mpTopology.NodeDeletionTimeout != nil {
+		g.Expect(mp.Spec.Template.Spec.NodeDeletionTimeout).To(Equal(mpTopology.NodeDeletionTimeout))
+	}
+
+	if mpTopology.NodeVolumeDetachTimeout != nil {
+		g.Expect(mp.Spec.Template.Spec.NodeVolumeDetachTimeout).To(Equal(mpTopology.NodeVolumeDetachTimeout))
+	}
+
+	if mpTopology.MinReadySeconds != nil {
+		g.Expect(mp.Spec.MinReadySeconds).To(Equal(mpTopology.MinReadySeconds))
+	}
+
+	if mpTopology.FailureDomains != nil && mp.Spec.Template.Spec.FailureDomain != nil {
```

Comment:
I think mp.Spec.Template.Spec.FailureDomain == nil is an error case if mpTopology.FailureDomains is not nil. Can we  drop this part from the if and check it with g.Expect inside the if branch?